gt/main.c

```diff
@@ -50,6 +65,78 @@ gt_setup_rss(struct gt_config *gt_conf)
 	return gatekeeper_setup_rss(port_in, gt_queues, gt_conf->num_lcores);
 }
 
+static void *
+get_ip_hdr(struct rte_mbuf *m, uint8_t *ip_version)
+{
+	struct ipv4_hdr *ip4_hdr = rte_pktmbuf_mtod(m, struct ipv4_hdr *);
+
+	*ip_version = (ip4_hdr->version_ihl & 0xF0) >> 4;
+	if (*ip_version != 4 && *ip_version != 6)
+		return NULL;
+
+	return ip4_hdr;
+}
+
+static int
+lookup_policy_decision(void *ip_hdr, uint8_t ip_version,
+	struct ggu_policy *policy, struct gt_instance *instance)
+{
+	struct gt_match_fields mf;
+
+	if (ip_version == 4) {
+		struct ipv4_hdr *ip4_hdr = (struct ipv4_hdr *)ip_hdr;
+
+		mf.proto = ETHER_TYPE_IPv4;
+		mf.ip.v4 = ip4_hdr->dst_addr;
+
+		if (ip4_hdr->next_proto_id == IPPROTO_TCP)
+			mf.dest_port = rte_be_to_cpu_16(
+				((struct tcp_hdr *)&ip4_hdr[1])->dst_port);
+		else if (ip4_hdr->next_proto_id == IPPROTO_UDP)
+			mf.dest_port = rte_be_to_cpu_16(
+				((struct udp_hdr *)&ip4_hdr[1])->dst_port);
+		else
+			return -1;
+
+		policy->flow.proto = ETHER_TYPE_IPv4;
+		policy->flow.f.v4.src = ip4_hdr->src_addr;
+		policy->flow.f.v4.dst = ip4_hdr->dst_addr;
+	} else {
+		struct ipv6_hdr *ip6_hdr = (struct ipv6_hdr *)ip_hdr;
+
+		mf.proto = ETHER_TYPE_IPv6;
+		rte_memcpy(mf.ip.v6, ip6_hdr->dst_addr, sizeof(mf.ip.v6));
+
+		if (ip6_hdr->proto == IPPROTO_TCP)
+			mf.dest_port = rte_be_to_cpu_16(
+				((struct tcp_hdr *)&ip6_hdr[1])->dst_port);
+		else if (ip6_hdr->proto == IPPROTO_UDP)
+			mf.dest_port = rte_be_to_cpu_16(
+				((struct udp_hdr *)&ip6_hdr[1])->dst_port);
+		else
+			return -1;
+
+		policy->flow.proto = ETHER_TYPE_IPv6;
+		rte_memcpy(policy->flow.f.v6.src, ip6_hdr->src_addr,
+			sizeof(policy->flow.f.v6.src));
+		rte_memcpy(policy->flow.f.v6.dst, ip6_hdr->dst_addr,
+			sizeof(policy->flow.f.v6.dst));
+	}
+
+	lua_getglobal(instance->lua_state, "lookup_policy");
+	lua_pushlightuserdata(instance->lua_state, &mf);
+	lua_pushlightuserdata(instance->lua_state, policy);
```

Comment:
The policy needs to know the source address of the outer IP header, that is, the IP address of the Gatekeeper server. This information is useful because a policy can map this IP address to the IX from where the packet is coming from in order to infer if the source address of the inner IP header is spoofed.

Moreover, parameter `struct gt_match_fields mf` is not a good idea, it's only adding the port number when it exists, all their other fields are already available in `policy`. A better idea is to pass the reference to the IP header, which the policy can properly cast to the correct IP header, and a pointer to the header above the inner IP header. This will enable policies to read any field of the header of packets.